pkg/manager/member/pvc_resizer.go

```diff
@@ -80,215 +79,361 @@ type pvcResizer struct {
+	ns := dc.Namespace
+
+	ctx := &componentVolumeContext{
+		comp:                  comp,
+		desiredVolumeQuantity: map[v1alpha1.StorageVolumeName]resource.Quantity{},
+	}
+
 	selector, err := label.NewDM().Instance(dc.GetInstanceName()).Selector()
 	if err != nil {
-		return err
+		return nil, err
 	}
-
-	// patch dm-master PVCs
-	{
-		pvcPrefix2Quantity := make(map[string]resource.Quantity)
+	switch comp {
+	case v1alpha1.DMMasterMemberType:
+		ctx.selector = selector.Add(*dmMasterRequirement)
+		if dc.Status.Master.Volumes == nil {
+			dc.Status.Master.Volumes = map[v1alpha1.StorageVolumeName]*v1alpha1.StorageVolumeStatus{}
+		}
+		ctx.sourceVolumeStatus = dc.Status.Master.Volumes
 		if quantity, err := resource.ParseQuantity(dc.Spec.Master.StorageSize); err == nil {
-			dmMasterMemberType := v1alpha1.DMMasterMemberType.String()
-			key := fmt.Sprintf("%s-%s-%s", dmMasterMemberType, dc.Name, dmMasterMemberType)
-			pvcPrefix2Quantity[key] = quantity
+			ctx.desiredVolumeQuantity[v1alpha1.GetStorageVolumeName("", v1alpha1.DMMasterMemberType)] = quantity
 		}
-		if err := p.patchPVCs(ns, selector.Add(*dmMasterRequirement), pvcPrefix2Quantity); err != nil {
-			return err
+	case v1alpha1.DMWorkerMemberType:
+		ctx.selector = selector.Add(*dmWorkerRequirement)
+		if dc.Status.Worker.Volumes == nil {
+			dc.Status.Worker.Volumes = map[v1alpha1.StorageVolumeName]*v1alpha1.StorageVolumeStatus{}
 		}
-	}
-
-	// patch dm-worker PVCs
-	if dc.Spec.Worker != nil {
-		pvcPrefix2Quantity := make(map[string]resource.Quantity)
+		ctx.sourceVolumeStatus = dc.Status.Worker.Volumes
 		if quantity, err := resource.ParseQuantity(dc.Spec.Worker.StorageSize); err == nil {
-			dmWorkerMemberType := v1alpha1.DMWorkerMemberType.String()
-			key := fmt.Sprintf("%s-%s-%s", dmWorkerMemberType, dc.Name, dmWorkerMemberType)
-			pvcPrefix2Quantity[key] = quantity
-		}
-		if err := p.patchPVCs(ns, selector.Add(*dmWorkerRequirement), pvcPrefix2Quantity); err != nil {
-			return err
+			ctx.desiredVolumeQuantity[v1alpha1.GetStorageVolumeName("", v1alpha1.DMWorkerMemberType)] = quantity
 		}
+	default:
+		return nil, fmt.Errorf("unsupported member type %s", comp)
 	}
-	return nil
-}
 
-func (p *pvcResizer) isVolumeExpansionSupported(storageClassName string) (bool, error) {
-	sc, err := p.deps.StorageClassLister.Get(storageClassName)
+	podVolumes, err := p.collectAcutalStatus(ns, ctx.selector)
 	if err != nil {
-		return false, err
+		return nil, err
 	}
-	if sc.AllowVolumeExpansion == nil {
-		return false, nil
-	}
-	return *sc.AllowVolumeExpansion, nil
+	ctx.actualPodVolumes = podVolumes
+
+	return ctx, nil
 }
 
-// patchPVCs patches PVCs filtered by selector and prefix.
-func (p *pvcResizer) patchPVCs(ns string, selector labels.Selector, pvcQuantityInSpec map[string]resource.Quantity) error {
-	if len(pvcQuantityInSpec) == 0 {
-		return nil
+// updateVolumeStatus build volume status from `actualPodVolumes` and call the `updateVolumeStatusFn` function.
```

Comment:
Where is the `updateVolumeStatusFn` function?

Reply:
obsolete comment, changed.